### DISCUSSION TITLE
TypeError: BigInt value can't be serialized in JSON

### DISCUSSION THREAD

**Discussion by sir-stevie** [paulmillr/micro-btc-signer#73] [Q&A]
I have created and signed a transaction using btc-signer library but when I try to broadcast it to blockstream test network I get the error, "TypeError: BigInt value can't be serialized in JSON" and I have no idea what to do at this point. I tried incorporating some custom method to convert the transaction object to a json object before broadcasting it but that didn't work, I got an error that said "sendrawtransaction RPC error: {"code":-22,"message":"TX decode failed. Make sure the tx has at least one input."}". The following is the relevant part of the code am using, 
"
// get inputs
    async getInputs(network, fromAddress, privateKey) {
        try {
            const newPrivateKey = base58.decode(privateKey).slice(1, 33);

            // Fetch utxos to be used as inputs for this transaction 
            const firstResponse = await axios.get(`${corsProxyUrl}/${blockstreamUrl}/address/${fromAddress}/utxo`);
            const utxos = firstResponse.data;

            const inputs = [];
            
            var inputCount = 0; // loop counter
            for (const element of utxos) {
                const utxo = {
                    amount: btc.Decimal.decode(`${element.value/100000000}`),
                    index: inputCount,
                    address: fromAddress,
                    txid: element.txid,
                    vout: element.vout,
                    witnessUtxo: {
                        amount: btc.Decimal.decode(`${element.value/100000000}`),
                        script: btc.p2wpkh(secp256k1.publicKeyCreate(newPrivateKey, true), network).script,
                    },
                };

                inputs.push(utxo);
                inputCount += 1;
            }

            return inputs;

        } catch(error) {
            this.handleError(error);
            console.log("An error occurred while getting inputs");
        }
    }

    // transaction to upload file
    async txToUploadData(network, data, fromAddress, toAddress, privateKey, amount){
        try {
            const newPrivateKey = base58.decode(privateKey).slice(1, 33);
            const inputs = await this.getInputs(network, fromAddress, privateKey);

            console.log("Done getting inputs... ", inputs);

            // append data as output
            const dataIn = Buffer.from(data);
            const outputScript = Buffer.concat([
                Buffer.from([btc.OP.PUSHDATA1, dataIn.length]),
                dataIn
            ]);

            console.log("Script: ", outputScript);

            const outputs = [
                {
                    address: toAddress, 
                    script: outputScript,
                    amount: 0n
                },
                {
                    address: toAddress,
                    amount: btc.Decimal.decode(`${amount}`)
                }
            ]

            console.log("Done setting outputs...");

            const selected = btc.selectUTXO(inputs, outputs, 'default', {
                changeAddress: fromAddress,
                feePerByte: 2n,
                bip69: true,
                createTx: true,
                network: network,
                allowUnknownOutputs: true
                
            });

            const {tx} = selected;
            tx.sign(newPrivateKey);
            tx.finalize();

            const headers = {
                "X-Requested-With": "XMLHttpRequest"
            }

            console.log("Transaction: ", JSON.stringify(tx));

            // broadcast the transaction
            const response = await axios({method: 'POST', url: `${corsProxyUrl}/${blockstreamUrl}/tx`, headers: headers, data: tx});

            // const response = axios.post(`${corsProxyUrl}/https://api.blockcypher.com/v1/bcy/test/txs/push`, JSON.stringify(pushtx))
            // .then(function(d) {console.log(d)});

            console.log("Done broadcasting transaction...", response.data);

            return tx;

        }catch(error) {
            console.log("An error occurred while sending 1st transaction");
            this.handleError(error);
        }
    }
    "

**Comment by victorkirov** [comment]
Try using tx.hex as the data field when making the request to blockstream

**Reply by sir-stevie** [reply]
I tried this and the program was able to create a transaction hex and took care of the "BigInt value can't be serialized in JSON" error but now am facing another problem. The endpoint API is responding with this message, "XHRPOST
http://127.0.0.1:8080/https://blockstream.info/testnet/api/tx
[HTTP/1.1 400 Bad Request 1630ms]

 
1

sendrawtransaction RPC error: {"code":-26,"message":"mandatory-script-verify-flag-failed (Operation not valid with the current stack size)"}

"

**Reply by victorkirov** [reply]
Your output script looks a bit strange. What are you trying to do exactly? 🤔

**Reply by sir-stevie** [reply]
Am trying to transfer bitcoin from one wallet to another and in the transaction am appending data read from a file. This is the flow:

1. First am fetching all the utxos from the sending wallet (done by the getInputs function)
2. Then I create an array of outputs (the file data and the bitcoin to transfer)

From there, I use the selectUTXO() method as directed in the documentation.

That's the flow of events in my code

**Reply by sir-stevie** [reply]
The error persists even after I exclude the output script from the code

**Reply by victorkirov** [reply]
Just took another look. I think your get utxos for address function is a bit off.

try:
```
async getInputs(network, fromAddress, privateKey) {
    try {
        const newPrivateKey = base58.decode(privateKey).slice(1, 33);

        // Fetch utxos to be used as inputs for this transaction 
        const firstResponse = await axios.get(`${corsProxyUrl}/${blockstreamUrl}/address/${fromAddress}/utxo`);
        const utxos = firstResponse.data;

        const inputs = [];
        
        for (const element of utxos) {
            const utxo = {
                index: element.vout,
                txid: element.txid,
                witnessUtxo: {
                    amount: element.value,
                    script: btc.p2wpkh(secp256k1.publicKeyCreate(newPrivateKey, true), network).script,
                },
            };

            inputs.push(utxo);
        }

        return inputs;
    } catch(error) {
        this.handleError(error);
        console.log("An error occurred while getting inputs");
    }
}

```

**Reply by sir-stevie** [reply]
The error persists, I will try debugging more see if I can come up with something
